pkg/atomicwriter/atomic_writer_test.go

```diff
@@ -0,0 +1,255 @@
+			path: "test_file.txt",
+			osFuncs: osFuncs{
+				remove: func(name string) error {
+					return os.ErrPermission
+				},
+				rename: func(oldpath, newpath string) error {
+					return os.ErrPermission
+				},
+				truncate: func(name string, size int64) error {
+					return os.ErrPermission
+				},
+				chmod: os.Chmod,
+			},
+			assert: func(buf *bytes.Buffer, wc io.WriteCloser, t *testing.T, err error) {
+				assert.Error(t, err)
+
+				assert.Contains(t, buf.String(), "Could not delete temporary file")
+				assert.Contains(t, buf.String(), "File may be left on disk")
+			},
+		},
+		{
+			name: "unable to chmod",
+			path: "test_file.txt",
+			osFuncs: osFuncs{
+				remove:   os.RemoveAll,
+				rename:   os.Rename,
+				truncate: os.Truncate,
+				chmod: func(name string, mode os.FileMode) error {
+					return os.ErrPermission
+				},
+			},
+			assert: func(buf *bytes.Buffer, wc io.WriteCloser, t *testing.T, err error) {
+				assert.NoError(t, err)
+				assert.Contains(t, buf.String(), "Could not set permissions on temporary file")
+
+				// Check that the file was still renamed
+				writer, ok := wc.(*atomicWriter)
+				assert.True(t, ok)
+				assert.NoFileExists(t, writer.tempFile.Name())
+				assert.FileExists(t, writer.path)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tmpDir, _ := ioutil.TempDir("", "atomicwriter")
+			defer os.RemoveAll(tmpDir)
+			path := filepath.Join(tmpDir, tc.path)
+
+			// Mock the logger output
+			buf := mockErrorLog()
+			defer unmockErrorLog()
+
+			writer, err := newAtomicWriter(path, 0644, tc.osFuncs)
+
+			if tc.errorOnCreate {
+				assert.Error(t, err)
+				tc.assert(buf, writer, t, err)
+				if writer != nil {
+					writer.Close()
+				}
+				return
+			}
+
+			assert.NoError(t, err)
+
+			// Try to write the file
+			_, err = writer.Write([]byte("test content"))
+			assert.NoError(t, err)
+
+			err = writer.Close()
+			tc.assert(buf, writer, t, err)
+		})
+	}
+}
+
+func TestDefaultDirectory(t *testing.T) {
+	writer, err := NewAtomicWriter("test_file.txt", 0644)
+	assert.NoError(t, err)
+	defer os.Remove("test_file.txt")
+
+	writer.Write([]byte("test content"))
+	err = writer.Close()
+	assert.NoError(t, err)
+	assert.FileExists(t, "./test_file.txt")
+}
+
+func writeFile(path string, permissions os.FileMode, content []byte) (err error, tempFilePath string) {
+	writer, err := NewAtomicWriter(path, permissions)
+	if err != nil {
+		return err, ""
+	}
+
+	tempFilePath = writer.(*atomicWriter).tempFile.Name()
+
+	_, err = writer.Write(content)
+	if err != nil {
+		return err, tempFilePath
+	}
+
+	return writer.Close(), tempFilePath
+}
+
+// Mocks the logger output to a buffer
+func mockErrorLog() *bytes.Buffer {
+	buf := &bytes.Buffer{}
+	log.ErrorLogger.SetOutput(buf)
+	return buf
+}
+
+// Unmocks the logger output, sending it back to os.Stderr
+func unmockErrorLog() {
+	log.ErrorLogger.SetOutput(os.Stderr)
+}
```

Comment:
In this case it doesn't seem that we have to worry about running tests in parallel but if we did I suspect the mocking and un-mocking might breakdown. To make testing the logs simpler I would have dependency injected the error logger function, much like you did the os funcs, since the atomic writer only uses `log.Error`.